### ISSUE TITLE
improve usage with Docker

### ISSUE THREAD

**Issue by mathroc** [mailpile/Mailpile#1698]
I'd like to improve the usage of mailpie from Docker. here is what I have in mind:

* [ ] fix Dockerfile so that it's at least able to build & run (#1697)
* [ ] use alpine to make the image much smaller (it's currently 441MB in #1697)
* [ ] fix gpg key creation (at least it failed on my one attempt to create a 4096 one)
* [ ] run mailepie with tor (inside the same image? or env vars link to an external tor node ? both ? etc.)
* [ ] expose web interface on port 80
* [ ] publish on hub.docker.com (auto build of master->latest and tags)
* [ ] add documentation about usage with docker (examples, volumes to mount, upgrade, etc.)

what do you think about this ? anything missing ?

**Comment by ryneeverett** [comment]
I'd add that there probably needs to be a distinction between development and production use cases in the documentation and likely the image/compose file as well.

**Comment by BjarniRunar** [comment]
Pinging @DoubleMalt - do you want to have an opinion on this?

**Comment by BjarniRunar** [comment]
This all sounds worth doing! Thank you!

My only caveat is, I like to request that you hold off on publishing to hub.docker.com until we go into release-candidate mode. We're iterating a bit too fast right now to want to open up to an audience that isn't willing to do development.

Regarding the GPG key generation, that is almost certainly due to a shortage of entropy and Mailpile timing out. I'm not sure how much you'll be able to do about that from the Docker environment, but if you have clever ideas then I'm all ears!

**Comment by mathroc** [comment]
@ryneeverett yes development probably have different needs than production mode. I don't have much knowledge about python dev so I'm mostly going on this for the production version. someone else might have better idea of how to improve developing with docker

@BjarniRunar don't worry I don't plan on publishing on the docker hub myself (I couldn't setup the proper hooks anyway)
ans finally about the GPG key generation. I found out that's it's because the version installed on ubuntu is gpg2 instead of gpg1.4. I'll have to look at how to install this version on the image until #1133 is resolved

**Comment by DoubleMalt** [comment]
I'm happy to join the fray and offer opinions. I'm just in the process of
getting our ARM docker hub ready for primetime ;)

**Comment by mathroc** [comment]
other things to consider:

* [ ] setup on first run so that it can write inside mounted volume
* [ ] setup variable env (language, home, password?, etc.)
* [ ] chown when starting (to fix permissions on mounted volumes)

**Comment by BjarniRunar** [comment]
Regarding the key generation timeout, this thread may help: https://stackoverflow.com/questions/26021181/not-enough-entropy-to-support-dev-random-in-docker-containers-running-in-boot2d - especially note the trick of mounting /dev/urandom as /dev/random inside the container. Most objections against /dev/urandom (preferring /dev/random) are obsolete, as far as I know, so I don't think I'm giving security-breaking advice here. ;-)

**Comment by mathroc** [comment]
@BjarniRunar the key generation failure comes of the upgrade to gnupg 2 in ubuntu 16.04. in #1699 I'm forcing gnupg 1. going back to 14.04 shoudl fix it. to remember to test I made a new PR #1706

**Comment by BjarniRunar** [comment]
OK. But you're still going to have the entropy problem. That's why Mailpile defaults to recommending 2048 bit keys - they're "secure enough" and we can actually successfully generate them much more reliably than 4k.

**Comment by mathroc** [comment]
it seems to be working fine when I'm using alpine, maybe it's because I'm running plain docker and not boot2docker

**Comment by rgarrigue** [comment]
Hi

I'm looking forward the official repo in hub.docker.com, any ETA in mind ?

Regards,

**Comment by almereyda** [comment]
When running `docker-compose up` with a current clone of the repository, I get:

```
Creating mailpile ... done
Attaching to mailpile
mailpile    | 
mailpile    | Required binary missing or unusable: GnuPG not found
mailpile    | 
mailpile    | If you know where it is, or would like to skip this test and run Mailpile
mailpile    | anyway, you can set one of the following environment variables:
mailpile    | 
mailpile    |     MAILPILE_GNUPG="/path/to/binary"
mailpile    | or
mailpile    |     MAILPILE_IGNORE_BINARIES="GnuPG"
mailpile    | 
mailpile    | Note that skipping a binary check may cause the app to become unstable or
mailpile    | fail in unexpected ways. If it breaks you get to keep both pieces!
mailpile    | 
mailpile exited with code 1
```

This means we need to add the environmental variable to the container with the current path to the GPG binary in Debian stretch. Adding

```
    environment:
      MAILPILE_GNUPG: "/usr/bin/gpg"
```

to the `mailpile:` service in `docker-compose.yml` allows for it to be run.

I also liked the discussion in https://community.mailpile.is/t/running-mailpile-from-a-docker-container/165 yet find the custom Docker wrapper scripts a little, say, unhandy.
